firedrake/preconditioners/base.py

```diff
@@ -109,25 +109,21 @@ def get_appctx(pc):
         return get_appctx(pc.getDM()).appctx
 
     @staticmethod
-    def new_snes_ctx(pc, op, bcs, mat_type, fcp=None, options_prefix=None):
+    def new_snes_ctx(pc, op, bcs, mat_type, fcp=None, options_prefix=None, pre_apply_bcs=True):
         """ Create a new SNES contex for nested preconditioning
         """
-        from firedrake.variational_solver import NonlinearVariationalProblem
+        from firedrake.variational_solver import LinearVariationalProblem
         from firedrake.function import Function
-        from firedrake.ufl_expr import action
         from firedrake.solving_utils import _SNESContext
 
         dm = pc.getDM()
         old_appctx = get_appctx(dm).appctx
         u = Function(op.arguments()[-1].function_space())
-        F = action(op, u)
+        L = 0
         if bcs:
-            bcs = tuple(bc._as_nonlinear_variational_problem_arg() for bc in bcs)
-        nprob = NonlinearVariationalProblem(F, u,
-                                            bcs=bcs,
-                                            J=op,
-                                            form_compiler_parameters=fcp)
-        return _SNESContext(nprob, mat_type, mat_type, old_appctx, options_prefix=options_prefix)
+            bcs = tuple(bc._as_nonlinear_variational_problem_arg(is_linear=True) for bc in bcs)
+        nprob = LinearVariationalProblem(op, L, u, bcs=bcs, form_compiler_parameters=fcp)
```

Comment:
This is the `PCSNESBase`, is it correct to create a linear VP rather than a nonlinear one, like it does currently?

Reply:
It looks like this changes from creating a `NonlinearVariationalProblem` with (implicitly) `action(op, u) == 0`, to a `LinearVariationalProblem` with `op == 0`. Is this a correct reading, and if so are the two identical in practice?

If we are assuming that `op` is always a bilinear form, then I think that should be added to the docstring. It isn't an obvious requirement for a `snes_context` given that a `snes` is for nonlinear problems.

Reply:
This is mathematically identical (and also at the UFL) to the old behavior, except that now we let LVP create the lifted residual `action(J, u)`. The only difference is that now we also advertise this as a linear problem.

Reply:
Ok thanks. In which case, please can you add a comment to the docstring and I'm happy for this to be resolved.

Reply:
As the docstring/comment mentions, we were always assuming that `op` is a bilinear form (hence the `action`) and that this method is only called by nested PC objects.

Reply:
The docstring for `new_snes_ctx` is:
```python
        """ Create a new SNES contex for nested preconditioning
        """
```
I'm not sure which comment you are referring to, but requirements on the arguments should be in the docstring anyway.